### ISSUE TITLE
Traverse asyncio.wait_for

### ISSUE THREAD

**Issue by dimaqq** [dimaqq/awaitwhat#5]
`asyncio.wait_for` waits for something with a timeout:
https://docs.python.org/3/library/asyncio-task.html#timeouts

we should traverse `wait_for` (i.e. show awaitable behind it), and report remaining timeout.

**Comment by Puneethnaik** [comment]
Greetings @dimaqq. New contributer here. I am really interested in the project and would like to contribute. Please guide me on the next steps.

**Comment by dimaqq** [comment]
Hi @Puneethnaik sorry for the delay on my end.

Today, each of the `asyncio.xxx` functions needs special, custom traversal.

Please check the code that handles `asyncio.sleep` and `asyncio.gather`:

Compare the implementation of `asyncio.sleep`:

https://github.com/python/cpython/blob/67e394562d67cbcd0ac8114e5439494e7645b8f5/Lib/asyncio/tasks.py#L649-L656
```py
    future = loop.create_future()
    h = loop.call_later(delay,
                        futures._set_result_unless_cancelled,
                        future, result)
    try:
        return await future
    finally:
        h.cancel()
```


With the code that handles `asynio.sleep`:

https://github.com/dimaqq/awaitwhat/blob/41254780191cb77a32fe420d697455bac8b14c64/awaitwhat/sleep.py#L15-L17

Roughly, this library uses the C module `awaitwhat/what.c` to reach the stack frame where the async task is "waiting"; then checks what function that frame belongs to; then tries to analyse that frame by looking at its local variables.

In this case, the implementation of `sleep` creates a blank future and schedules a "callback" that will wake up this future. The calling task then waits of this future. At any given time, only task may be running, the rest may be ready or waiting for something. If a task is waiting for `sleep`, then `await future` did not return yet, and (probably) future is still pending. (I'm not sure off the top of my head what happens if future already had its result set, but event loop is busy with other tasks).

Thus, the job of this library is to examine the circumstances of that pending future. It does so by looking at local variable `future` (the future) and `h` the handler for the scheduled "callback" pulls out something interesting about it, in this case `when()` will the callback be fired.

The rest is handling corner cases.

How did I arrive at this?

My method was:
* construct some test code, that gets a test task into the waiting state
* run `ipython` (jupyter-notebook may also work)
* load this test code, run it to get a task in the waiting state
   * this was a bit tricky because `ipython` too uses the event loop 🙈 
* then inspect all tasks and find the target task (or inspect this very task directly)
* get a reference to it's "last" frame
* poke around interactively at that frame, look up the source code, and try to find something useful to hold on to
* when interactive code works, drop it into `awaitwhat/smth.py`:
   * a `mine` function which detects if this new submodule should process the given "last" frame
   * and `decode` function that pulls out something useful from that frame
* write a few unit tests and catch corner cases

Note: this project currently does not work on cpython 3.10; it does on 3.8 and 3.9 though.

**Comment by Puneethnaik** [comment]
@dimaqq no worries :slightly_smiling_face: Thanks a lot for the detailed introduction to the code. I will start following your directions, read the code repository and try to solve the task.

**Comment by Puneethnaik** [comment]
Greetings @dimaqq . I have been working on this issue for a while. I am able to show a vertex for wait_for denoting the timeout remaining and show the awaitable behind it. For example, if the code is 
```
import asyncio
from examples.test_shield import frob_a_branch
import awaitwhat


FUTURES = []


async def main():
    t = asyncio.create_task(test())
    await do_work()
    await t


async def do_work():
    loop = asyncio.get_running_loop()
    f = loop.create_future()
    FUTURES.append(f)
    branch = frob_a_branch(f)
    await asyncio.wait_for(branch, 15)

async def frob_a_branch(f):
    await f



def name(t):
    try:
        # Python3.8
        return t.get_name()
    except AttributeError:
        return f"Task-{id(t)}"


async def test():
    import sys
    await asyncio.sleep(1)
    try:
        tt = asyncio.all_tasks()
        print(awaitwhat.dot.dumps(tt))
    finally:
        for f in FUTURES:
            f.set_result(None)


if __name__ == "__main__":
    asyncio.run(main())
```
The code produces the following graph
![graph6](https://user-images.githubusercontent.com/13706737/125322255-33788e80-e35b-11eb-939f-5e0283105aab.png)
Is this fine @dimaqq ? your expert opinion will be extremely valuable for me in evaluating the code written by me. After your approval,  i will start working on the corner test cases and writing proper tests. Thanks in advance :slightly_smiling_face:.

**Comment by dimaqq** [comment]
Looks good! Please post a pull request 🙏🏿

**Comment by Puneethnaik** [comment]
@dimaqq Thanks for the feedback!  Here is the pull request: https://github.com/dimaqq/awaitwhat/pull/31

**Comment by dimaqq** [comment]
Thank you @Puneethnaik❗

**Comment by dimaqq** [comment]
Released in `awaitwhat==21.1` 🚀

**Comment by Puneethnaik** [comment]
Thanks a lot for accepting my contribution @dimaqq. It is an honor to contribute to such a wonderful project !
